slides/02_traits.md

```diff
@@ -0,0 +1,376 @@
+
+
+class Child(HasStrictTraits):
+    age = Int
+    father = Instance(Parent)
+
+    def _age_changed(self, old, new):
+        print('Age changed from %s to %s ' % (old, new))
+
+    @observe('father.last_name')
+    def _dad_name_updated(self, event):
+        print('DAD name', self.father.last_name)
+
+```
+
+```python
+dad = Parent(last_name='Zubizaretta')
+c = Child(father=Parent)
+```
+
+```python
+dad.last_name = 'Valderrama'
+```
+
+```python
+def handler(event):
+    print("handler", event.object, event.name, event.old, event.new)
+```
+
+```python
+c = Child(father=Parent(last_name='Ram'))
+c.observe(handler, 'father, age')
+```
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Trait change notification
+
+- Static: `def _<trait_name>_changed()`
+- Decorator: `@observe('extended.trait.name')`
+- Dynamic:
+
+```obj.observe(handler, 'extended.trait.name')
+```
+
+- See documentation: https://docs.enthought.com/traits/traits_user_manual/notification.html
+
+<!-- #endregion -->
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Exercise
+
+- Modify the first example to produce the above example
+- Add a `first_name` trait
+- Add a `Bool` trait to specify if person is alive
+- Add an `Enum` for the gender of the child
+
+
+<!-- #endregion -->
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Solution
+
+<!-- #endregion -->
+
+```python
+from traits.api import Bool, Enum
+
+class Parent(HasStrictTraits):
+    last_name = Str('')
+
+
+class Child(HasStrictTraits):
+    age = Int
+    father = Instance(Parent)
+    first_name = Str('')
+    alive = Bool(True)
+    gender = Enum('female', 'male', 'neither')
+
+    def _age_changed(self, old, new):
+        print('Age changed from %s to %s ' % (old, new))
+
+    @observe('father.last_name')
+    def _dad_name_updated(self, event):
+        print('Dad name', self.father.last_name)
+
+```
+
+```python
+p = Parent(last_name='Ray')
+c = Child(age=21, father=p, first_name='Romano', gender='male')
+```
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Setting default values
+
+- For simple cases, use the default of the trait
+- For more complex cases use a special method
+- A simple example
+
+<!-- #endregion -->
+
+```python
+import datetime
+
+from traits.api import HasStrictTraits, Date, Range
+
+class Thing(HasStrictTraits):
+    date = Date()
+    age = Int(12)
+
+    def _date_default(self):
+        print('default')
+        return datetime.datetime.today()
+```
+
+
+```python
+t = Thing()
+```
+
+```python
+type(c.age)
+```
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Trait Lists
+
+
+<!-- #endregion -->
+
+```python
+from traits.api import List
+
+class Bowl(HasStrictTraits):
+    fruits = List(Str)
+
+    def _fruits_changed(self, o, n):
+        print("Fruits changed", o, n)
+
+```
+
+```python
+b = Bowl()
+b.fruits = ['apple']
+b.fruits.append('mango')
+```
+
+<!-- #region slideshow={"slide_type": "slide"} -->
+## Trait List events
```

Comment:
This isn't very relevant for our tutorial. I compiled a list of `traits` concepts we'll need for each stage:
- Stage 2:
    - Array, cached_property, Dict, File, HasStrictTraits, List, Property
- Stage 4:
    - file_browser.py
        - Directory, Event
    - `_trait_default` factory method
- Stage 5.1:
    - `observe("list.items")`
    - `HasStrictTraits.__init__`
- Stage 5.2:
    - Button
- Stage 5.3;
    - Bool, Enum

Reply:
Thank you for this list @siddhantwahal !! We should make sure what we introduce in each section supports all future stages...

Reply:
Thanks for the extensive comments. I have introduced list events because you do listen to them in stage 5.1 right?  The way I envisaged these initial slides was to give users a feel with the basic ideas. When the stage-specific slides are added they simply need to discuss the new traits that are being introduced either directly in the context of the code or with a simpler example. My goal is to first introduce everyone to the basic ideas with toy examples. Nothing more.  The subsequent slides should expand on the additional traits. Thoughts?  This makes it much easier to just introduce the stage, the stage specific traits in a couple of slides and then show a demo.